tensorboard/loader.py

```diff
@@ -0,0 +1,413 @@
+      be greater than or equal to the offset field of any Record.
+
+    Raises:
+      IOError: If the stat call failed or file shrunk.
+
+    :rtype: int
+    """
+    with self._lock:
+      if self._should_stat():
+        self._stat()
+      return self._size
+
+  def get_next_record(self):
+    """Reads one record.
+
+    When this method is first called, it will spawn the thread and
+    block until a record is read. Once the thread starts, it will queue
+    up records which can be read without blocking. The exception is
+    when we reach the end of the file, in which case each repeated call
+    will be synchronous. There is no background polling. If new data is
+    appended to the file, new records won't be buffered until this
+    method is invoked again. The caller should take care to meter calls
+    to this method once it reaches the end of file, lest they impact
+    performance.
+
+    Returns:
+      A Record object, or None if there are no more records available
+      at the moment.
+
+    Raises:
+      IOError: If this instance has been closed.
+      tf.errors.DataLossError: If corruption was encountered in the
+          records file.
+      Exception: To propagate any exceptions that may have been thrown
+          by the read operation in the other thread. If an exception is
+          thrown, then all subsequent calls to this method will rethrow
+          that same exception.
+
+    :rtype: Record
+    """
+    with self._lock:
+      if self._is_closed:
+        raise IOError('%s is closed' % self)
+      if not self._thread.is_alive():
+        self._thread.start()
+      else:
+        record = self._get_record()
+        if record is not None:
+          return record
+        self._has_reached_end = False
+        self._wake_up_producer.notify()
+      while (self._read_exception is None and
+             not self._has_reached_end and
+             not self._records):
+        self._wake_up_consumers.wait()
+      return self._get_record()
+
+  def close(self):
+    """Closes event log reader if open.
+
+    If the i/o thread is running, this method blocks until it has been
+    shut down.
+
+    Further reads are not permitted after this method is called.
+
+    Raises:
+      Exception: To propagate any exceptions that may have been thrown
+          by the close operation in the other thread. If an exception
+          is thrown, then all subsequent calls to this method will
+          rethrow that same exception.
+    """
+    with self._lock:
+      if not self._is_closed:
+        self._is_closed = True
+        if not self._thread.is_alive():
+          self._reader = None
+          return
+        self._wake_up_producer.notify()
+      while self._reader is not None:
+        self._wake_up_consumers.wait()
+      if self._close_exception is not None:
+        raise self._close_exception
+
+  def _get_record(self):
+    if self._read_exception is not None:
+      raise self._read_exception
```

Comment:
An unfortunate quirk of Python is that this loses the original stack trace (unlike in Java, in Python errors do not retain a stack trace, and rather the `raise` mechanism takes the current stack trace).

If you want to retain the stack trace, you should store in `self._read_exception` not the `Exception` object but the 3-tuple result of `sys.exc_info`. Then, if you only care about Python 2, here instead of `raise self._read_exception` write `(type, value, traceback) = self._read_exception; raise type, value, traceback`. [Docs.](https://docs.python.org/2/reference/simple_stmts.html#the-raise-statement) Since we want to be hybrid, instead write `six.reraise(*self._read_exception)`. [Docs.](https://pythonhosted.org/six/#six.reraise) (Caveat scriptor: I've not used `six.reraise` before, but it seems like the right thing.)

If you think that this is too much trouble, that's your call.

(This likewise above with `raise self._close_exception`.)

By the way, thanks for being careful about raising exceptions in the right threads. That's a good thing to have.